### DISCUSSION TITLE
[Feature request] Global/root form error, independent of form fields

### DISCUSSION THREAD

**Discussion by bennettdams** [TanStack/form#716] [Ideas] [answered]
Hey!
I think it would be nice to have a global/root form error that you can show next to the form.

Imagine a user submits its form, (client) validation is successful, request is sent to the backend. But something goes wrong (network, business logic, etc.), so we'll e.g. get a 400 response. 
For such, I want to show a message next to the form. 

You could obviously do this yourself, but then you also have to keep track of the form's submissions, as the error may goes away after the next submit. So integrating it as part of the form would be nice.

The easiest solution in my head is to have a new reserved field on the `FormApi`, something like `rootError`. One should have the possibility to get and set the value and it should be reset on a submission.

If I understand correctly, [some API-related changes](https://github.com/TanStack/form/discussions/623) will soon allow to set field errors from the `onSubmit` callback. But this feature request is something different.

**Comment by depsimon** [comment]
Not sure if it has been added since your post, currently you can have an error that is bound to the form (not to a specific field).

```tsx
// Set the errors
form.setErrorMap({
  onSubmit: formErrors.join(","),
});
```

```tsx
// Show the errors
<p>{form.state.errors.join(', ')}</p>
```

**Comment by crutchcorn** [comment, marked as answer]
We've actually supported form-level errors for a LONG time. Anything set from useForm's validators will be attached to the form object, not fields

**Reply by bennettdams** [reply]
Thanks for the reply, but it was not obvious to me how to manually set an error on root level (and not on the field level).
The answer from @depsimon pushed me in the right direction though! 

Is this the canonical way to do this?

```tsx
  const form = useForm({
    defaultValues: {
      firstName: '',
      lastName: '',
    },
    onSubmit: async ({ value, formApi }) => {
      // imagine a fetch call here
      // const response = update(value)
      const isError = true;

      if (isError) formApi.setErrorMap({ onSubmit: 'Something went wrong' });
    },
  });
```

**Reply by Balastrong** [reply]
No, [this example](https://tanstack.com/form/latest/docs/framework/react/examples/field-errors-from-form-validators) is the recommended way to set errors.

That onSubmit in your code is the function triggered _after_ the form has been succesfully submitted. `useForm` has a `validators` object you can use to define functions, their return value will be automatically set into form (or fields) errors.

```ts
const form = useForm({
    defaultValues: {
      age: 0,
    },
   // Fired on successful submit
    onSubmit: async ({ value }) => {
      console.log(value)
    },
    validators: {
      // Add validators to the form the same way you would add them to a field
      onChange({ value }) {
        if (value.age < 13) {
          return 'Must be 13 or older to sign'
        }
        return undefined
      },
    },
})
```

More on the [Form Validation docs](https://tanstack.com/form/latest/docs/framework/react/guides/validation)

**Reply by bennettdams** [reply]
@Balastrong I think there is a misunderstanding? My use case is:
"All fields are valid, I submitted the form and use the valid fields to send a network request."
Imagine this network request is failing for whatever reason (maybe a failed validation on the backend).
When the error response is there, I want to set a "root" error for the form, as I can't assign this kind of error to a specific field.

**Reply by depsimon** [reply]
FYI we are having kind of the same discussion here: https://github.com/TanStack/form/issues/1105

**Reply by Balastrong** [reply]
You can set a `validators.onSubmitAsync` function that does the network request and if something goes wrong it returns a string. That string will be automatically set as global form error.

You can see that in the example I linked at line 40, setting "invalid data" at form level.

Is that what you were looking for?

**Reply by bennettdams** [reply]
>in the example I linked at line 40

Oh, that is perfect! I guess I was confused because the title says "Field errors", but this type of error is usually named "root error".

In my defense, I'm pretty sure this haven't existed yet when I created the discussion.

Thank you!

**Reply by johanfive** [reply]
It is very nice yes 🙂 but 🙃 ...
```js
validators: {
    onSubmitAsync: async ({ value }) => {}
}
```
should not be used as a **substitute** for `onSubmit`...
```js
onSubmit: ({ value: formData }) => {
    return createUser.mutateAsync(formData)
        .then(form.reset)
        .catch((non2xxApiResponse) => {
            // set form level error from here
        });
    },
```
`validators.onSubmitAsync` is beautiful if you're working with a server API that has an endpoint dedicated to validating data, but realistically, how often does that happen IRL?
The example provided implements a `verifyAgeOnServer` which is a good example to grasp the idea, but in practice we'd almost always just POST the data to the endpoint and trust that it'll respond with a 400 and some details in the response payload.

What the OP (and myself now) are hoping for is a function that does what
```
return {
    // The `form` key is optional
    form: 'Invalid data',
    fields: {}
};
```
does.

(And perhaps that "form" key would not be optional in this context. The form key **is** the main point of the feature request)

(Big fan though! ❤️ )

**Reply by crutchcorn** [reply]
@johanfive I fundamentally do not understand what you're saying. That's exactly how onSubmit (both sync and async) work today.

**Reply by johanfive** [reply]
I am saying there is a difference between the `onSubmit` property and the `validators` property in the object we pass as argument to `useForm()`.

I am saying `onSubmit` **!==** `validators.onSubmit`. Right?

```js
const UserCreationForm = () => {
  const createUser = useCreateUserMutation();
  const form = useForm<CreateUserPayload>({
    defaultValues: {
      name: "",
      age: 0,
      email: "",
    },
    onSubmit: ({ value: formData }) => {
      return createUser.mutateAsync(formData)
        .then(form.reset)
        .catch((e) => {
            // here we would deal with a 409 Conflict (email already exists) response from the API
            // because we would NOT first do a "checkEmail" request inside the validators.onSubmit,
            // we'd just fire the "createUser" request and hope for the best.
        });
    },
    validators: {
        onSubmitAsync: async ({ value }) => {
            // Would NOT call some "checkEmail" function here.
            // Because the API does not have an endpoint for this.
        }
    }
  });
  return (
    <form onSubmit={(e) => {
      e.preventDefault();
      e.stopPropagation();
      void form.handleSubmit();
    }}>
    // ...
    </form>
  );
};
```

Again it's entirely possible I'm missing something, but I guess that's an indicator the docs could use some tweaking then?

**Reply by crutchcorn** [reply]
Oh, okay. I see now. It was not sufficiently clear from your last message that you meant `onSubmit` as in not the validator.

It should be extremely trivial for us to add a non-breaking API for this akin to:

```tsx
formApi.addErrors({
  onChange: {
    fields: {
      firstName: "Test"
    }
  }
})
```

Can you make an issue with your _complete_ code sample and this code snippet suggestion from me so we can track as a feature request outside of discussions?

**Reply by johanfive** [reply]
I can create a ticket and I will if you still believe it's necessary after this last comment:

After some trial and error I realized we can achieve what I described earlier, by doing something like:
```js
    onSubmit: ({ value: formData }) => {
      return createUser.mutateAsync(formData)
        .then(form.reset)
        .catch((non2xxApiResponse) => {
          // THIS works, but is this _the way_?
          form.setErrorMap({
            onSubmit: "Email already taken (or whatever)"
          });
        });
    },
    validators: {
        onSubmitAsync: async ({ value }) => {
            // do other kinds of "pre-flight" validations here
            // This is still an awesome feature
        }
    }
```
It works but I suspect it _might_ not be a good idea to mess with the `errorMap`'s `onSubmit` directly like that 🤔 
You're the expert you tell us 😄

**Reply by crutchcorn** [reply]
That's fine to do so. I was thinking that we could even support setErrorMap with the `fields` property but after reconsidering probably not needed.

**Reply by johanfive** [reply]
@crutchcorn I mean it would actually be nice to have the option to do:
```js
    onSubmit: ({ value: formData }) => {
        return createUser.mutateAsync(formData)
            .then(form.reset)
            .catch(() => {
                form.setErrorMap({
                    onSubmit: {
                        form: "Form submission failed",
                        fields: {
                            email: "Already taken"
                        }
                    }
                });
        });
    },
    validators: {
        onSubmitAsync: async ({ value }) => { ... }
    }
```
Especially since we can see that the type of `state.errorMap.onSubmit` is
```js
(property) onSubmit?: string | false | {
    form?: ValidationError;
    fields: Partial<Record<string, ValidationError>>;
} | null | undefined
```
in a `<form.Subscribe selector={(state) => [state.errorMap.onSubmit]}></form.Subscribe>`

(An added benefit of all this is that if implemented and documented, you're then really encouraging people to think about their error handling)

**Reply by gregorbg** [reply]
I really want to second the idea expressed in the previous post by @johanfive. Such an API would be a blessing, because it seems that `formApi.setErrorMap({ onSubmit: ... })` can only set *one* global error message and not field-level errors.

The argument about the return value of `onSubmit` already (effectively) allowing such a setter is very compelling to me, and it would tremendously help me in my codebase to adopt this wonderful library.

**Reply by johanfive** [reply]
@gregorbg I've opened a [feature request here](https://github.com/TanStack/form/discussions/1141).

Feel free to upvote and tell your friends 😄

**Reply by MurzNN** [reply]
```ts
          form.setErrorMap({
            onSubmit: "Email already taken (or whatever)"
          });
```
How to make this type of a global error to be shown in the form area? On my form, it just doesn't appear anywhere, only per-field errors appear. Maybe I need to place a special component to the form layout to make it shown?
